rfcs/OIP-001-unified-urn.md

```diff
@@ -0,0 +1,47 @@
+
+The goal of this RFC is to decide what is a good and persistent URN format that can be used across our tools.
+
+## Background
+
+### Current formats
+
+To understand the needs of this initiation better, let's take a look at how each of our tools generate their URN format.
+
+| Resource            | Format                                       | Example                                            |
+| :------             | :----                                        | :-----                                             |
+| Meteor's RDBMS      | `{service}::{host}/{database}/{table}`       | `postgres::10.283.86.19:5432/user_db/user_role`    |
+| Meteor's BigQuery   | `{service}::{project}/{dataset}/{table}`     | `bigquery::odpf-prod/datamart/daily_booking`       |
+| Meteor's Metabase   | `{service}::{host}/dashboards/{dashboardID}` | `metabase::my-metabase-server.com/dashboards/872`  |
+| Shield's Resource   | `{resource_type}/{namespace}/{resource_id}`  | `r/namespace-id/resource-name`                     |
+| Guardian's BigQuery | `{resource_id}`                              | `metabase:293`                                     |
+
+There are few things that we can improve here:
+
+- Using `{host}` as part of the URN will damage persistency (mostly on meteor's). Resource location should be allowed to change without causing its generated URN to be invalid. The reason behind using `{host}` as URN component is for removing ambiguity on URN.
+- Using `/` as a separator has a few issues:
+    1. When passing a resource URN as route parameter via `http` protocol, this URN will need to be encoded.
+    2. Even if it is encoded, some **services** or **proxies** may not be able to `route-match` properly. (e.g [gorillamux](https://github.com/gorilla/mux/issues/639) default behaviour)
+
+### Limited resource referencing between tools
+
+Instead of each tools defining their own URN formats, it will be better, if possible, all tools or services have the same urn format when talking about a resource or asset.
+
+Since different tools are using different format, this would prevent resource referencing (or potentially sharing?) between tools without helps from an extra mapping layer (either by service or library).
+
+## Requirements
+
+Our final unified URN should handle these cases:
+1. Persist through change of resource location. (e.g. DB is moved to another server)
+2. Can easily be used on URL without relying on services to handle the encoding/decoding.
+3. Should be globally unique, or at least within an organization (AKAB / DKAB).
+4. SAMPLE CASE: If we somehow have two different Metabases, we should be able to differentiate which metabase it is from URN without relying on `host`.
+
+And these are cases that are great to have:
+1. -
+
+## Proposal
```

Comment:
I highly recommend we stick to the IETF standard definition of URN from [RFC8141](https://datatracker.ietf.org/doc/html/rfc8141) (even if it's only a subset). 

[RFC 8141: Section 2](https://datatracker.ietf.org/doc/html/rfc8141#section-2) defines the syntax for URNs.

NSS, r & q-components are allowed to have `/` character in the URN but, for simplicity we can recommend not to have it. 

1. For all ODPF products, we can use `odpf` (or `ODPF`) as the [Namespace Identifier (NID)](https://datatracker.ietf.org/doc/html/rfc8141#section-2.1).
2. Every ODPF product should use the product name as [Namespace Specific String](https://datatracker.ietf.org/doc/html/rfc8141#section-2.2). For example, all resources managed by Entropy would have `entropy` as the NSS. 

NID and NSS combined forms the `assigned-name`: `urn:<NID>:<NSS>` --> `urn:odpf:entropy`. This assigned-name uniquely identifies every product within odpf.

Optional components (which are defined by the entity that owns the NSS) can be appended to `assigned-name` to form resource-level identifiers. 

Optional components can have some generic restrictions that we follow: For example all optional components following the namespace should match the pattern `^[A-Za-z0-9-]+$`.  

With all these combined: The URN for  a "resource"  of kind "firehose" in project "foo" with name "f1" managed by "entropy" will be `urn:odpf:entropy:foo:firehose:f1`

Reply:
> With all these combined: The URN for a "resource" of kind "firehose" in project "foo" with name "f1" managed by "entropy" will be urn:odpf:entropy:foo:firehose:f1

@spy16  CMIIW this would mean, that the same firehose resource will have different urns depending where it is being stored. 🤔 

Shouldn't a resource (e.g. a firehose) have a same urn regardless where it is stored?
We could do `urn:odpf:firehose:foo:f1` (assuming we are keeping `urn:odpf`). This way all services can refer to the same firehose.

Reply:
Yes. this is one concern we need to address that urn can stay consistent across the services.

Reply:
@StewartJingga It actually depends what do you mean by "same". Are 2 firehoses, both in project "foo", 1 deployed using entropy and other deployed using something else same? Because if we ever have that scenario, both firehose instances would have urn as `urn:odpf:firehose:foo:f1`.  

If we know for sure that we don't have to worry about such ambiguities, then it should be okay to not use the product name as NSS and use the "resource kind" instead.

Reply:
> Are 2 firehoses, both in project "foo", 1 deployed using entropy and other deployed using something else same? Because if we ever have that scenario, both firehose instances would have urn as urn:odpf:firehose:foo:f1

@spy16 when you say 2 firehoses, is it the same firehose but different instance? If it is the same firehose resource (not instance), then it should only have one, but if it a different firehose (different name) then it should have different urn. But I am not talking about the deployment here, more about the firehose resource itself.